### ISSUE TITLE
Improve initialization of embedded python interpreter

### ISSUE THREAD

**Issue by dalboris** [vgc/vgc#236]
See #58 for more details.

Notes:
- We only support Python 3.6+, and we don't support python virtual_envs for now (just the python installation distributed with python), this simplifies things, compared to, say: https://github.com/FreeCAD/FreeCAD/blob/1995f9d0bac63820c5c42ac0075c91a49cbad119/src/Base/Interpreter.cpp#L512
- Since we embed Python without the intent of executing a single Python script, it makes sense for us to call [PySys_SetArgvEx](https://docs.python.org/3.8/c-api/init.html#c.PySys_SetArgvEx) with `updatepath=0`, see documentation:
> Note It is recommended that applications embedding the Python interpreter for purposes other than executing a single script pass 0 as updatepath, and update sys.path themselves if desired.
- Starting with Python 3.7, there is no need to call [https://docs.python.org/3.8/c-api/init.html#c.PyEval_InitThreads](PyEval_InitThreads):
> Changed in version 3.7: This function is now called by Py_Initialize(), so you don’t have to call it yourself anymore.

**Comment by dalboris** [comment]
Other note: pybind11 has minimal support for embedding Python (cf. https://pybind11.readthedocs.io/en/stable/advanced/embedding.html), but it doesn't allow to pass arguments using PySys_SetArgvEx (neither sub-interpreters, but we don't use this for now). The role of PySys_SetArgvEx isn't very important in our use case (just allowing python scripts to know command line arguments passed to VGC via sys.argv, not sure if we will ever need this), but since it's quite easy to support it, we direcly use the Python C API rather than pybind11::scoped_interpreter. Also, in the future, this will make it more flexible to support sub-interpreter and have finer control for multi-threading.

**Comment by dalboris** [comment]
Note: there seems to be quite a bit of work happening recently to improve the init sequence of CPython, especially since Python 3.8:

https://www.python.org/dev/peps/pep-0432/
https://www.python.org/dev/peps/pep-0587/
https://docs.python.org/3/c-api/init_config.html

Also, I was kinda confused as to the role of Py_SetProgramName(), and seems that this may have been part of the confusion:
https://bugs.python.org/issue31349 

In conclusion, maybe we'll have to revisit in a few years which functions of the C API we call to perform Python initialization, and hopefully, things will actually be cleaner / more flexible.

**Comment by dalboris** [comment]
Alright, after more research, looking at the source code, and re-reading more carefully the documentation, it looks like Py_SetProgramName() is in fact quite useless. In particular, see the following:

https://docs.python.org/3.8/c-api/intro.html

> On most systems (in particular, on Unix and Windows, although the details are slightly different), Py_Initialize() calculates the module search path based upon its best guess for the location of the standard Python interpreter executable, assuming that the Python library is found in a fixed location relative to the Python interpreter executable. In particular, it looks for a directory named lib/pythonX.Y relative to the parent directory where the executable named python is found on the shell command search path (the environment variable PATH).
> 
> For instance, if the Python executable is found in /usr/local/bin/python, it will assume that the libraries are in /usr/local/lib/pythonX.Y. (In fact, this particular path is also the “fallback” location, used when no executable file named python is found along PATH.) The user can override this behavior by setting the environment variable PYTHONHOME, or insert additional directories in front of the standard path by setting PYTHONPATH.
> 
> The embedding application can steer the search by calling Py_SetProgramName(file) before calling Py_Initialize(). Note that PYTHONHOME still overrides this and PYTHONPATH is still inserted in front of the standard path. An application that requires total control has to provide its own implementation of Py_GetPath(), Py_GetPrefix(), Py_GetExecPrefix(), and Py_GetProgramFullPath() (all defined in Modules/getpath.c).

More specifically, it seems that:
- If you don't call `Py_SetProgramName()` at all, then Python will search for a binary called `python3` in the PATH, and look for Python libs relative to this using black magic.
- If you call `Py_SetProgramName(name)` with `name` not containing any slash, then Python will search for a binary called `name` in the PATH, and look for Python libs relative to this using black magic.
- If you call `Py_SetProgramName(path)` with `path` containing at least one slash, then it will resolve the path using the current working directory, and look for Python libs relative to this using black magic.
- If you call `Py_SetPythonHome(home)`, then it overrides any previous call of `Py_SetProgramName()`, which are then useless. In fact, it seems that `Py_SetProgramName()` isn't even used to populate `sys.executable`, which is magically populated using other means (this has to be confirmed with more tests on all platforms).

The moral of the story is that `Py_SetProgramName(argv[0])` is definitely not reliable in the general case. It might help if Python is located nearby vgcillustration and if `argv[0]` is an absolute path (or is just the program name but is in the PATH), but it might even make things WORSE if our desired Python is system-installed (i.e., would have been found without calling  `Py_SetProgramName()`) and vgcillustration is located far from this system-installed Python. I have in fact noticed this on my local build: the first time I added `Py_SetProgramName(argv[0])` (without using `Py_SetPythonHome()` yet), then it failed to find Python, while everything was working using the black magic, probably because it was finding `python3` in my PATH, which is pretty scary to be honest... this might not at all be the Python manually specified to CMake.

For more control, we can use `Py_SetPythonHome()`, which makes things much cleaner in my opinion. And if we use `Py_SetPythonHome()`, then we don't need `Py_SetProgramName()`, although it cannot hurt to set it.

**Comment by dalboris** [comment]
Useful source files to look at for more details on the mechanism (search for "program" in these files):

- Black magic for Unix: https://github.com/python/cpython/blob/3.8/Modules/getpath.c
- Black magic for Windows: https://github.com/python/cpython/blob/3.8/PC/getpathp.c
- Setting PythonHome and ProgramName:
  https://github.com/python/cpython/blob/3.8/Include/cpython/initconfig.h
  https://github.com/python/cpython/blob/3.8/Python/initconfig.cpp
  https://github.com/python/cpython/blob/3.8/Python/pathconfig.c
